apis/vertexai/v1alpha1/featurestore_types.go

```diff
@@ -0,0 +1,196 @@
+	// +required
+	KMSKeyRef *refsv1beta1.KMSCryptoKeyRef `json:"kmsKeyRef,omitempty"`
+}
+
+// VertexAIFeaturestoreSpec defines the desired state of VertexAIFeaturestore
+// +kcc:proto=google.cloud.aiplatform.v1beta1.Featurestore
+type VertexAIFeaturestoreSpec struct {
+	// The Project that this resource belongs to.
+	// +required
+	ProjectRef *refsv1beta1.ProjectRef `json:"projectRef"`
+
+	// The location for the resource.
+	// +required
+	Location string `json:"location"`
+
+	// The VertexAIFeaturestore name. If not given, the metadata.name will be used.
+	ResourceID *string `json:"resourceID,omitempty"`
+
+	// Optional. Used to perform consistent read-modify-write updates. If not set,
+	//  a blind "overwrite" update happens.
+	// +kcc:proto:field=google.cloud.aiplatform.v1beta1.Featurestore.etag
+	Etag *string `json:"etag,omitempty"`
+
+	// Optional. The labels with user-defined metadata to organize your
+	//  Featurestore.
+	//
+	//  Label keys and values can be no longer than 64 characters
+	//  (Unicode codepoints), can only contain lowercase letters, numeric
+	//  characters, underscores and dashes. International characters are allowed.
+	//
+	//  See https://goo.gl/xmQnxf for more information on and examples of labels.
+	//  No more than 64 user labels can be associated with one Featurestore(System
+	//  labels are excluded)."
+	//  System reserved label keys are prefixed with "aiplatform.googleapis.com/"
+	//  and are immutable.
+	// +kcc:proto:field=google.cloud.aiplatform.v1beta1.Featurestore.labels
+	Labels map[string]string `json:"labels,omitempty"`
+
+	// Optional. Config for online storage resources. The field should not
+	//  co-exist with the field of `OnlineStoreReplicationConfig`. If both of it
+	//  and OnlineStoreReplicationConfig are unset, the feature store will not have
+	//  an online store and cannot be used for online serving.
+	// +kcc:proto:field=google.cloud.aiplatform.v1beta1.Featurestore.online_serving_config
+	OnlineServingConfig *Featurestore_OnlineServingConfig `json:"onlineServingConfig,omitempty"`
+
+	// Optional. TTL in days for feature values that will be stored in online
+	//  serving storage. The Feature Store online storage periodically removes
+	//  obsolete feature values older than `online_storage_ttl_days` since the
+	//  feature generation time. Note that `online_storage_ttl_days` should be less
+	//  than or equal to `offline_storage_ttl_days` for each EntityType under a
+	//  featurestore. If not set, default to 4000 days
+	// +kcc:proto:field=google.cloud.aiplatform.v1beta1.Featurestore.online_storage_ttl_days
+	OnlineStorageTtlDays *int32 `json:"onlineStorageTtlDays,omitempty"`
+
+	// Optional. Customer-managed encryption key spec for data storage. If set,
+	//  both of the online and offline data storage will be secured by this key.
+	// +kcc:proto:field=google.cloud.aiplatform.v1beta1.Featurestore.encryption_spec
+	EncryptionSpec *EncryptionSpec `json:"encryptionSpec,omitempty"`
```

Comment:
does this support other encryption format?

Reply:
Not sure I fully understand the question, do mean what happens if `EncryptionSpec` is not set? If unset, a "Google-owned and Google-managed encryption key" will be used as the default encryption. This is pretty standard across the GCP resources.

Reply:
sorry for the confusion, I should add the comment in `type EncryptionSpec struct`. It looks like a one-of and currently only support KMS key. I'm wondering if it supports other secret management approaches.

Reply:
I did not see any oneof? It is the same struct shared by many VertexAI resources.